src/core.ts

```diff
@@ -188,6 +189,26 @@ export function onTrigger(key: string): void {
   }
 }
 
+export function onCustomLog(kvps: { [key: string]: any }): void {
+  if (state === State.Activated) {
+    const event: IEventData = {
+      type: CustomEventName,
+      state: kvps,
+      time: getTimestamp(),
+    };
+    addEvent(event);
+  }
+}
+
+export function onSetPageInfo(pageId: string, userId: string): void {
+  // only allow setting pageId and userId if Clarity isn't currently running
+  if (state === State.Loaded || state === State.Unloaded) {
+    cid = userId;
+    setClarityCookie(userId);
+    impressionId = pageId;
```

Comment:
>pageId [](start = 19, length = 6)

Both of these should set the value only if parameter is non-null/defined (to allow setting only one of them).